dotnet/src/Connectors/Connectors.Memory.AzureCosmosDBMongoDB/AzureCosmosDBMongoDBVectorStoreRecordCollection.cs

```diff
@@ -0,0 +1,351 @@
+        }
+    }
+
+    #region private
+
+    private async Task CreateIndexAsync(string collectionName, CancellationToken cancellationToken)
+    {
+        var indexCursor = await this._mongoCollection.Indexes.ListAsync(cancellationToken).ConfigureAwait(false);
+        var indexes = indexCursor.ToList(cancellationToken).Select(index => index["name"].ToString()) ?? [];
+        var uniqueIndexes = new HashSet<string?>(indexes);
+
+        var indexArray = new BsonArray();
+
+        // Create separate index for each vector property
+        foreach (var property in this._vectorStoreRecordDefinition.Properties.OfType<VectorStoreRecordVectorProperty>())
+        {
+            // Use index name same as vector property name with underscore
+            var indexName = $"{this._storagePropertyNames[property.PropertyName]}_";
+
+            // If index already exists, proceed to the next vector property
+            if (uniqueIndexes.Contains(indexName))
+            {
+                continue;
+            }
+
+            // Otherwise, create a new index
+            var searchOptions = new BsonDocument
+            {
+                { "kind", property.IndexKind },
+                { "numLists", this._options.NumLists },
+                { "similarity", property.DistanceFunction },
+                { "dimensions", property.Dimensions }
+            };
+
+            if (this._options.EfConstruction is not null)
+            {
+                searchOptions["efConstruction"] = this._options.EfConstruction;
+            }
+
+            var indexDocument = new BsonDocument
+            {
+                ["name"] = indexName,
+                ["key"] = new BsonDocument { [property.PropertyName] = "cosmosSearch" },
+                ["cosmosSearchOptions"] = searchOptions
+            };
+
+            indexArray.Add(indexDocument);
+        }
+
+        if (indexArray.Count > 0)
+        {
+            var createIndexCommand = new BsonDocument
+            {
+                { "createIndexes", collectionName },
+                { "indexes", indexArray }
+            };
+
+            await this._mongoDatabase.RunCommandAsync<BsonDocument>(createIndexCommand, cancellationToken: cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    private FilterDefinition<BsonDocument> GetFilterById(string id)
+        => Builders<BsonDocument>.Filter.Eq(document => document[AzureCosmosDBMongoDBConstants.MongoReservedKeyPropertyName], id);
+
+    private FilterDefinition<BsonDocument> GetFilterByIds(IEnumerable<string> ids)
+        => Builders<BsonDocument>.Filter.In(document => document[AzureCosmosDBMongoDBConstants.MongoReservedKeyPropertyName].AsString, ids);
+
+    private async Task<bool> InternalCollectionExistsAsync(CancellationToken cancellationToken)
+    {
+        using var cursor = await this._mongoDatabase.ListCollectionNamesAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
```

Comment:
Did you try some of the suggestions here, where you supply a filter with the name to `ListCollectionNamesAsync` to avoid enumerating all names?

https://stackoverflow.com/questions/25017219/how-to-check-if-collection-exists-in-mongodb-using-c-sharp-driver

Reply:
Good point, thanks!